### PR TITLE
Purecap kernel pmap

### PR DESCRIPTION
**Pull request by qwattash**

### CODE REVIEW

File: sys/riscv/riscv/vm_machdep.c [hunk 1]
```diff
@@ -39,13 +39,13 @@ __FBSDID("$FreeBSD$");
 #include <sys/systm.h>
 #include <sys/limits.h>
 #include <sys/proc.h>
-#include <sys/sf_buf.h>
```
[the remark this answers is not in the document]
Comment (qwattash) [reply sys/riscv/riscv/vm_machdep.c:42]: I dropped this. I'm not sure it should even be in the purecap kernel branch at all.

File: sys/amd64/amd64/pmap.c [hunk 2]
```diff
@@ -7146,7 +7145,7 @@ pmap_enter_quick_locked(pmap_t pmap, vm_offset_t va, vm_page_t m,
 {
 	pt_entry_t newpte, *pte, PG_V;
 
-	KASSERT(va < kmi.clean_sva || va >= kmi.clean_eva ||
+	KASSERT(!VA_IS_CLEANMAP(va) ||
```
Comment (bsdjhb) [inline sys/amd64/amd64/pmap.c:7148]: I really should pull this out and upstream it to FreeBSD.

File: sys/amd64/amd64/pmap.c [hunk 3]
```diff
@@ -11450,3 +11449,13 @@ DB_SHOW_COMMAND(ptpages, pmap_ptpages)
 	}
 }
 #endif
+
```
Comment (bsdjhb) [inline sys/amd64/amd64/pmap.c:11452]: These blank lines should be trimmed.

File: sys/riscv/riscv/pmap.c [hunk 4]
```diff
@@ -566,7 +581,7 @@ pmap_bootstrap_l3(vm_offset_t l1pt, vm_offset_t va, vm_offset_t l3_start)
 	}
 
 	/* Clean the L2 page table */
-	memset((void *)l3_start, 0, l3pt - l3_start);
+	memset((void *)l3_start, 0, (vaddr_t)l3pt - (vaddr_t)l3_start);
```
Comment (bsdjhb) [inline sys/riscv/riscv/pmap.c:584]: We should replace 'vaddr_t' with 'ptraddr_t' when bringing these changes into dev.

Comment (qwattash) [reply sys/riscv/riscv/pmap.c:584]: Indeed, I missed this. I thought `vaddr_t` had permanently been removed and would fail to build?

Comment (brooksdavis) [reply sys/riscv/riscv/pmap.c:584]: I've added `ptraddr_t`, but not widely adopted it to avoid churn during the current merges.

Comment (qwattash) [reply sys/riscv/riscv/pmap.c:584]: Makes sense, I'll keep an eye out for `vaddr_t` then, as I was assuming the compiler would catch it.

**Comment by brooksdavis** [comment]
The riscv64 build error needs:
```diff
index b4454132374f..fb7dfdbbd698 100644
--- a/sys/riscv/riscv/pmap.c
+++ b/sys/riscv/riscv/pmap.c
@@ -160,9 +160,7 @@ __FBSDID("$FreeBSD$");
 #include <machine/pcb.h>
 #include <machine/sbi.h>
 
-#if __has_feature(capabilities)
 #include <cheri/cheric.h>
-#endif
 
 #define        NUL1E           (Ln_ENTRIES * Ln_ENTRIES)
 #define        NUL2E           (Ln_ENTRIES * NUL1E)
```

File: sys/amd64/amd64/pmap.c (hunk 2 above)
Comment (qwattash) [reply sys/amd64/amd64/pmap.c:7148]: Do you want me to split this out so you can cherry-pick or leave it for now?

Comment (qwattash) [reply sys/amd64/amd64/pmap.c:7148]: I split the commit, so hopefully upstreaming it is easier.

**Review by bsdjhb — APPROVED** [review]
The RISC-V bits and history all look ok to me.

**Pull request merged**